pkg/util/socket.go

```diff
@@ -0,0 +1,214 @@
+// After prevCtxCancelFunc function corresponding to last timer executed, prevCtx and prevCtxCancelFunc will be reassigned
+// in order to represent brand-new timer.
+func modifyLoglevel(newLogLevel string) error {
+	mutex.Lock()
+	defer mutex.Unlock()
+
+	// Change klog log level to new value
+	var loglevel klog.Level
+	if err := loglevel.Set(newLogLevel); err != nil {
+		return err
+	}
+	currentLogLevel = newLogLevel
+
+	// Cancel the previous timer.
+	if prevCtxCancelFunc != nil {
+		prevCtxCancelFunc()
+	}
+	prevCtx, prevCtxCancelFunc = context.WithCancel(context.Background())
+	return nil
+}
+
+// reset creates a timer to make klog recover to start-up log level.
+func reset(ctx context.Context, duration time.Duration) {
+	defer runtime.HandleCrash()
+	select {
+	// Create a timer
+	case <-time.After(duration):
+		var loglevel klog.Level
+		mutex.Lock()
+		defer mutex.Unlock()
+		if err := loglevel.Set(startupLogLevel); err != nil {
+			return
+		}
+		currentLogLevel = startupLogLevel
+		klog.InfoS("Klog recover to start-up log level successfully", "startupLogLevel", startupLogLevel)
+	// Cancel previous timer
+	case <-ctx.Done():
+		klog.InfoS("Cancel previous timer successfully")
+	}
+}
+
+// installKlogLogLevelHandler registers the HTTP request patterns that can set/get current klog log level
+func installKlogLogLevelHandler(mux *http.ServeMux, startup string) {
+	currentLogLevel, startupLogLevel = startup, startup
+	// Register the HTTP request patterns that can change klog log level
+	mux.Handle(setLogLevelPath, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		values := r.URL.Query()
+		rawLevel := values.Get("level")
+		rawDuration := values.Get("duration")
+		// Escape the data that needs to be output to the log, Prevent Reflected cross-site scripting
+		rawLevel = html.EscapeString(rawLevel)
+		rawDuration = html.EscapeString(rawDuration)
+
+		if rawDuration == "" {
+			rawDuration = DefaultDuration
+		}
+		var duration time.Duration
+		var err error
+		// Validate argument in request
+		if level, err := strconv.ParseInt(rawLevel, 10, 64); err != nil || level <= 0 {
+			responseError(&w, exampleSocketCli, http.StatusBadRequest)
+			return
+		}
+		if duration, err = time.ParseDuration(rawDuration); err != nil || duration.Milliseconds() <= 0 {
+			responseError(&w, exampleSocketCli, http.StatusBadRequest)
+			return
+		}
+
+		if err := modifyLoglevel(rawLevel); err != nil {
+			responseError(&w, fmt.Sprintf("Failed to change klog log level. Error: %v\n", err.Error()), http.StatusInternalServerError)
+			return
+		}
+
+		mutex.RLock()
+		// Create a timer to make klog recover to start-up log level.
+		// There will be more than one timer using same prevCtx variable under extreme conditions.
+		// Therefore, put reset function in mutex range.
+		go reset(prevCtx, duration)
+		responseOk(&w, fmt.Sprintf("Change klog log level to %s successfully and  for %v\n", currentLogLevel, duration))
+		mutex.RUnlock()
```

Comment:
Duding the read lock,  there is also an write lock requirement in gorutime, is that as your expect?

Reply:
I didn't understand what you meant. Can you provide a more detailed scenario?

Reply:
I mean that you have locked with a readlock, and then start a goruntine in which a write lock is required.

Reply:
The wirte lock is in func `modifyLoglevel`,using to protect `prevCtx`.
```
func modifyLoglevel(newLogLevel string) error {
	mutex.Lock()
	defer mutex.Unlock()

	// Change klog log level to new value
	var loglevel klog.Level
	if err := loglevel.Set(newLogLevel); err != nil {
		return err
	}
	currentLogLevel = newLogLevel

	// Cancel the previous timer.
	if prevCtxCancelFunc != nil {
		prevCtxCancelFunc()
	}
	prevCtx, prevCtxCancelFunc = context.WithCancel(context.Background())
	return nil
}
```

Reply:
There is also write lock in the reset gorutine function. In the scenario previous timer if up and select case goes into the write lock require, but at the same time you are doing modifyLoglevel and have acquired the lock. Now the modifyLoglevel runs the cacel func of previous context. What will happen?

Reply:
You considered the scenario so carefully👍
if modifyLoglevel get the wirte lock first,modifyLoglevel will set klog and the reset will set klog after, it is **abnormal.**
if reset get the write lock first ,it will reset klog and the modifyLoglevel  will set klog ,it will be normal.

This usage scenario is to temporarily increase the klog level during debugging without considering concurrency, so there should be no need to deal with the low-probability events you mentioned.
what do you think？
@lowang-bh